students/jbutts/Lesson04/mailroom2.py

```diff
@@ -0,0 +1,196 @@
+            #  to terminal
+        else:
+            response = str(input("Donor {} doesn't exist, add them (y/n)?:".format(donor_name)))
+            if response.lower() == 'y':
+                add_contribution(donor_name, amount, donor_db)
+                print(format_email(donor_name, amount, sum(donor_db[donor_name])))
+    return donor_db
+
+
+def format_email(donor, amount, total_contribution):
+    '''
+    Create a thank you email for a donor.
+    '''
+
+    mail_str = ''.join((
+        "Dear {},\n\n",
+        "Thank you for your recent contribution of ${:.2f}.\n\n",
+        "We appreciate your generosity in support of our mission.\n\n",
+        "Thank you for your lifetime contributions of ${:.2f}.\n\n"
+        "Warmest Regards,\n\n",
+        "Charity Staff\n")).format(donor, amount, total_contribution)
+
+    return mail_str
+
+
+def add_contribution(donor, amount, donor_db):
+    '''
+    After record_contribution() prompts user for input, add records to dictionary as needed.
+    '''
+
+    if donor_db.get(donor):
+        if isinstance(donor_db[donor], list):
+            if isinstance(amount, list):
+                donor_db[donor].extend(amount)  # when amount is a list
+            else:
+                donor_db[donor].extend([amount])  # in case we get passed a float for amount
+        else:
+            donor_db[donor].append(amount)  # add new to dict
+        # otherwise, add a single contribution with a new user
+    else:
+        if isinstance(amount, list):
+            donor_db[donor] = amount
+        else:
+            donor_db[donor] = [amount]
+    return donor_db
+
+
+def send_letter_to_all(donor_db):
+    '''
+    Write text files containing letter contents for all donors in the database
+    '''
+
+    for key, value in donor_db.items():
+        tempfilepath = tempfile.gettempdir() + "/" + str(key.replace(' ', "_")) + ".txt"
+        with open(tempfilepath, "w") as outputfile:
+            print("writing letter and storing in: " + tempfilepath)
+            outputfile.writelines(format_email(key, value[len(value)-1], sum(donor_db[key])))
+            outputfile.close()
+
+
+def switcher(arg, donor_db):
+    '''
+    Use a switcher dictionary as a way to present a menu-driven interface for users
+    '''
+
+    # Use a dictionary as a switch statement!
+    switcher = {
+        1: record_contribution,
+        2: show_donors,
+        3: send_letter_to_all,
+        4: exit,
+    }
+    func = switcher.get(arg, lambda: print("Invalid Entry"))
+    if func == exit:
+        func()
+    else:
+        func(donor_db)
+
+
+def main(donor_db):
+    '''
+    main loop, continually present users with menu options
+    '''
+
+    while True:
+        switcher(int(input(print_menu())), donor_db)
+
+
+if __name__ == '__main__':
+    # We're not getting imported, run main():
+    DONOR_DB = {}  # dict that contains user/donation records
```

Comment:
all caps are used for constants - donor_db isn't a constant so should be all lower case